lib/ProductOpener/Nutrition.pm

```diff
@@ -308,20 +373,1471 @@ sub convert_nutrient_to_100g ($nutrient_ref, $original_per, $original_per_quanti
+		{
+			my $input_set_index = -1;
+			foreach my $input_set_ref (@{$request_body_ref->{product}{nutrition}{input_sets}}) {
+
+				$input_set_index++;
+
+				my $ignore_set = 0;
+
+				if (ref($input_set_ref) ne 'HASH') {
+					add_error(
+						$response_ref,
+						{
+							message => {id => "invalid_type_must_be_object"},
+							field => {id => "nutrition.inputs_sets[$input_set_index]"},
+							impact => {id => "input_set_ignored"},
+						},
+						200
+					);
+					next;
+				}
+
+				my $source = $input_set_ref->{source};
+				my $preparation = $input_set_ref->{preparation};
+				my $per = $input_set_ref->{per};
+
+				if ((not defined $source) or ($source eq "")) {
+					add_error(
+						$response_ref,
+						{
+							message => {id => "missing_field"},
+							field => {id => "nutrition.inputs_sets[$input_set_index].source"},
+							impact => {id => "input_set_ignored"},
+						},
+						200
+					);
+					$ignore_set = 1;
+				}
+				if ((not defined $preparation) or ($preparation eq "")) {
+					add_error(
+						$response_ref,
+						{
+							message => {id => "missing_field"},
+							field => {id => "nutrition.inputs_sets[$input_set_index].preparation"},
+							impact => {id => "input_set_ignored"},
+						},
+						200
+					);
+					$ignore_set = 1;
+				}
+				if ((not defined $per) or ($per eq "")) {
+					add_error(
+						$response_ref,
+						{
+							message => {id => "missing_field"},
+							field => {id => "nutrition.inputs_sets[$input_set_index].per"},
+							impact => {id => "input_set_ignored"},
+						},
+						200
+					);
+					$ignore_set = 1;
+				}
+
+				if ($ignore_set) {
+					next;
+				}
+
+				# If we are passed unspecified_nutrients, set them in the input set
+				# If unspecified_nutrients is undef, we delete the field
+				if (exists $input_set_ref->{unspecified_nutrients}) {
+
+					# If unspecified_nutrients exists but is undef, we delete the field
+					if (not defined $input_set_ref->{unspecified_nutrients}) {
+						if (exists $input_sets_hash_ref->{$source}{$preparation}{$per}{unspecified_nutrients}) {
+							delete $input_sets_hash_ref->{$source}{$preparation}{$per}{unspecified_nutrients};
+						}
+					}
+					elsif (ref($input_set_ref->{unspecified_nutrients}) eq 'ARRAY') {
+						# We only keep valid nutrient ids
+						my @unspecified_nutrients = ();
+						foreach my $nid (@{$input_set_ref->{unspecified_nutrients}}) {
+							if (exists $valid_nutrients{$nid}) {
+								push @unspecified_nutrients, $nid;
+							}
+							else {
+								add_error(
+									$response_ref,
+									{
+										message => {id => "unknown_nutrient"},
+										field => {
+											id => "nutrition.inputs_sets[$input_set_index].unspecified_nutrients",
+											value => $nid
+										},
+										impact => {id => "nutrient_ignored"},
+									},
+									200
+								);
+							}
+						}
+						if (scalar(@unspecified_nutrients) > 0) {
```

Comment:
Potential error: what if it was not empty and it is now… you have to remove the eventual old "unspecified_nutrients" key.

Reply:
Well that's exactly what the else clause below does.